File: coin_change.cpp
// Time Complexity: O(m+n) = O(n) where m is the amount and n is the number of coins
// Space Complexity: O(m * n)
// Did this code successfully run on Leetcode : Yes

// Approach: Dynamic Programming
// For each coin we have two choices, either choose or not choose
// If we choose this coin, value will be 1 + dp[i][j - coins[i - 1]]
// If we do not choose this coin, value will be dp[i - 1][j]

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int m = coins.size(), n = amount;
        vector<vector<int>> dp(m+1, vector<int>(n+1));
        for(int j = 1; j <=n; j++){
            dp[0][j] = amount+1;
        }
        for(int i = 1; i <= m; i++){
            for(int j = 1; j <= n; j++){
                //If coin we have is greater than amount
                if(coins[i - 1] > j){
                    dp[i][j] = dp[i - 1][j];
                }
                else{
                    dp[i][j] = min(dp[i - 1][j], 1 + dp[i][j - coins[i - 1]]);
                }
            }
        }
        int ans = dp[m][n];
        if(ans >= amount+1){return -1;}
        return ans;
    }
};

File: CoinChange.cs
// time complexity - O(mn) where m= number of coins, m = amount
// space complexity - O(n) where n = amount
public class Solution
{
    public int CoinChange(int[] coins, int amount)
    {
        int n = amount;
        int m = coins.Length;
        int[] matrix = new int[n + 1];
        matrix[0] = 0;
        for (int j = 1; j <= n; j++)
        {
            matrix[j] = 99999;
        }
        for (int i = 1; i <= m; i++)
        {
            for (int j = 1; j <= n; j++)
            {
                // to optimiza space complexity
                if (coins[i - 1] <= j)
                {
                    matrix[j] = Math.Min(matrix[j], 1 + matrix[j - coins[i - 1]]);
                }
                //till denominaton of coin< amount
                // if(coins[i-1]>j)
                // {
                //     // oth case
                //     matrix[i,j] = matrix[i-1,j];
                // }
                // else
                // {
                //     matrix[i,j] = Math.Min(matrix[i-1,j], 1+ matrix[i,j-coins[i-1]]);
                // }               
            }
        }
        int re = matrix[n];
        if (re >= amount + 1)
        {
            return -1;
        }
        return re;
    }
}